cstate/powermgr_cstate_tests.sh

```diff
@@ -0,0 +1,322 @@
+    local unexpected_cstate=0.00
+
+    columns="sysfs,CPU%c1,CPU%c6"
+    turbostat_output=$(turbostat -i 10 --quiet \
+        --show $columns sleep 10 2>&1)
+    test_print_trc "Turbostat log: $turbostat_output"
+    all_deepest_cstate=$(echo "$turbostat_output" |
+        awk '{for(i=0;++i<=NF;)a[i]=a[i]?a[i] FS $i:$i} END{for(i=0;i++<=NF;)print a[i]}' | grep "CPU%c6")
+    test_print_trc "The deepest core cstate is: $all_deepest_cstate"
+    if [[ $all_deepest_cstate =~ $unexpected_cstate ]]; then
+        die "CPU core did not enter the deepest cstate!"
+    else
+        test_print_trc "All the CPU enter the deepest cstate!"
+    fi
+}
+
+#The Core C6 is only supported on Intel® Server platform
+#This funtion is to check Core C6 residency during runtime
+judge_cc6_residency_during_idle() {
+    columns="Core,CPU%c1,CPU%c6"
+    turbostat_output=$(turbostat -i 10 --quiet \
+        --show $columns sleep 10 2>&1)
+    test_print_trc "Turbostat log: $turbostat_output"
+    CC6_val=$(echo "$turbostat_output" | grep -E "^-" | awk '{print $3}')
+    test_print_trc "CPU Core C6 residency :$CC6_val"
+    [[ -n $CC6_val ]] || die "CPU Core C6 residency is not available."
+
+    # Judge whether CC6 residency is available during idle
+    turbostat_CC6_value=$(echo "scale=2; $CC6_val > 0.00" | bc)
+    if [[ $turbostat_CC6_value -eq 1 ]]; then
+        test_print_trc "CPU Core C6 residency is available \
+during idle when $current_cpuidle_driver is running"
+    else
+        die "Did not get CPU Core C6 residency during idle \
+when $current_cpuidle_driver is running"
+    fi
+}
+
+test_cpu_core_c7_residency_intel_idle() {
+    check_intel_idle
+    judge_cc7_residency_during_idle
+}
+
+test_cpu_core_c7_residency_intel_s2idle() {
+    check_intel_idle
+    judge_cc7_residency_during_s2idle
+}
+
+test_cpu_core_c6_residency_intel_idle() {
+    check_intel_idle
+    judge_cc6_residency_during_idle
+}
+
+cc_state_disable_enable() {
+    local cc=$1
+    local setting=$2
+
+    for ((i = 0; i < cpu_num; i++)); do
+        #Find Core Cx state
+        cc_num=$(grep . /sys/devices/system/cpu/cpu0/cpuidle/state*/name |
+            sed -n "/$cc$/p" | awk -F "/" '{print $8}' | cut -c 6)
+        test_print_trc "Core $cc state name is: $cc_num"
+        [[ -n "$cc_num" ]] || block_test "Did not get Core $cc state."
+        #Change Core Cx state
+        do_cmd "echo $setting > /sys/devices/system/cpu/cpu$i/cpuidle/state$cc_num/disable"
+        let deeper=$cc_num+1
+        #Change deeper Core Cx state
+        for ((j = deeper; j < state_num; j++)); do
+            do_cmd "echo $setting > /sys/devices/system/cpu/cpu$i/cpuidle/state$j/disable"
+        done
+    done
+}
+
+disable_cc_check_pc() {
+    local cc=$1
+    local pc_y=$2
+    local pc_n=$3
+    local cpu_num
+    local columns
+
+    cpu_num=$(lscpu | grep "^CPU(s)" | awk '{print $2}')
+    state_num=$(ls "${CPU_BUS_SYSFS_PATH}"/cpu0/cpuidle | grep -c state)
```

Comment:
No need {} if you already use "":
`state_num=$(ls "$CPU_BUS_SYSFS_PATH"/cpu0/cpuidle | grep -c state)`

Reply:
Thanks, updated!